presto-main/src/main/resources/webapp/assets/cluster.js

```diff
@@ -0,0 +1,740 @@
+                        </div>
+                    </div>
+                    <div className="col-xs-4">
+                        <div className="stat-title">
+                            <span className="text" data-toggle="tooltip" data-placement="right" title="Total amount of memory reserved by all running queries">
+                                Reserved Memory (B)
+                            </span>
+                        </div>
+                    </div>
+                    <div className="col-xs-4">
+                        <div className="stat-title">
+                            <span className="text" data-toggle="tooltip" data-placement="right" title="Moving average of CPU time utilized per second">
+                                Parallelism
+                            </span>
+                        </div>
+                    </div>
+                </div>
+                <div className="row stat-line-end">
+                    <div className="col-xs-4">
+                        <div className="stat stat-large">
+                            <span className="stat-text">
+                                {this.state.blockedQueries[this.state.blockedQueries.length - 1]}
+                            </span>
+                            <span className="sparkline" id="blocked-queries-sparkline"><div className="loader">Loading ...</div></span>
+                        </div>
+                    </div>
+                    <div className="col-xs-4">
+                        <div className="stat stat-large">
+                            <span className="stat-text">
+                                {formatDataSizeBytes(this.state.reservedMemory[this.state.reservedMemory.length - 1])}
+                            </span>
+                            <span className="sparkline" id="reserved-memory-sparkline"><div className="loader">Loading ...</div></span>
+                        </div>
+                    </div>
+                    <div className="col-xs-4">
+                        <div className="stat stat-large">
+                            <span className="stat-text">
+                                {formatCount(this.state.cpuTimeRate[this.state.cpuTimeRate.length - 1])}
+                            </span>
+                            <span className="sparkline" id="cpu-time-rate-sparkline"><div className="loader">Loading ...</div></span>
+                        </div>
+                    </div>
+                </div>
+            </div>
+        </div>);
+    }
+});
+
+ReactDOM.render(
+        <ClusterHUD />,
+        document.getElementById('cluster-hud')
+);
+
+var Query = React.createClass({
+    render: function()
+    {
+        var query = this.props.query;
+        var summary = query.state;
+        var progress = 0;
+
+        var completedDrivers = query.completedDrivers;
+        var runningDrivers = query.runningDrivers;
+        var queuedDrivers = query.queuedDrivers;
```

Comment:
I would guess it should be possible to precompute all variables required for rendering and pass them to template engine. Is it possible in React?